### PR TITLE
[CARBONDATA-2792][schema restructure] Create external table fails post schema restructure.

### PR DESCRIPTION
**Pull request by mohammadshahidkhan**
**Problem**
Once the table schema is restructured "(_do column drop and column add)_".
The api  **`org.apache.carbondata.spark.util.CarbonSparkUtil.getRawSchema(carbonRelation: CarbonRelation) : String`**  to get visible columns raw schema string in ascending order of their ordinal value throws ArrayIndexOutOfBoundException while creating external table.
The api prepares the array  of raw column schema of visible columns in the ascending  order of schema ordinal.It uses the schemaOrdinal as index to prepare the array, so the colum having the schemaOrdinal value more than the visible column count will cause the ArrayIndexOutOfBoundException.
**Solution**
1. Filter the visible and valid columns
2.  sort the columns based on the schema ordinal.
3. Prepare the raw column schema based on the sorted columns index.

Be sure to do all of the following checklist to help us incorporate 
your contribution quickly and easily:

 - [X] Any interfaces changed?
 NA
 - [X] Any backward compatibility impacted?
 NA
 - [X] Document update required?
NA
 - [X] Testing done
        Please provide details on 
        - Whether new unit test cases have been added or why no new tests are required?
        - How it is tested? Please attach test report.
        - Is it a performance related change? Please attach the performance test report.
        - Any additional information to help reviewers in testing this change.
       added test case to validate the create external table post restructure scenario.
 - [X] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA. 
NA

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/6311/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7569/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/6029/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/6037/

**Comment by mohammadshahidkhan** [comment]
retest sdv please

**Comment by brijoobopanna** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7637/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/6363/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/6379/

File: integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/createTable/TestCreateExternalTable.scala [hunk 1]
```diff
@@ -111,5 +115,13 @@ class TestCreateExternalTable extends QueryTest with BeforeAndAfterAll {
     }
     assert(exception.getMessage().contains("Create external table as select"))
   }
+  test("create external table with post schema resturcture") {
```
[the remark this answers is not in the document]
Comment (mohammadshahidkhan) [reply integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/createTable/TestCreateExternalTable.scala]: @KanakaKumar Thanks for review.
 I have modified the test case to validate the data post restructure and external table creation.
Please review once.

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7662/

**Comment by KanakaKumar** [comment]
LGTM

File: integration/spark2/src/main/scala/org/apache/carbondata/spark/util/CarbonSparkUtil.scala [hunk 2]
```diff
@@ -87,18 +88,20 @@ object CarbonSparkUtil {
     val fields = new Array[String](
       carbonRelation.dimensionsAttr.size + carbonRelation.measureAttr.size)
     val carbonTable = carbonRelation.carbonTable
+    val columnSchemas: mutable.Buffer[ColumnSchema] = carbonTable.getTableInfo.getFactTable.
+      getListOfColumns.asScala.filter(_.getSchemaOrdinal != -1).filter(!_.isInvisible).
```
Comment (kunal642) [inline integration/spark2/src/main/scala/org/apache/carbondata/spark/util/CarbonSparkUtil.scala]: both filter conditions can be combined to one like .filter(a>2 && b<10).

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/6393/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7681/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7686/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/6082/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/6412/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/6084/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/6418/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7692/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7721/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/6447/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/6114/

**Comment by kunal642** [comment]
LGTM

**Pull request closed without merging**